man/k-version-compare.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.4.
.TH K-VERSION-COMPARE-STATIC "1" "May 2020" "koalephant-shell-script-library 2.8.1" "User Commands"
.SH NAME
k-version-compare-static,  - Compare two version strings
.SH SYNOPSIS
.B k-version-compare-static
[\fI\,OPTION\/\fR]... \fI\,versionString1 versionString2 \/\fR[\fI\,operator=gte\/\fR]
.SH DESCRIPTION
Compare two version strings
.PP
Valid operators are eq/= (equal), neq/!= (not equal), gt/> (greater than), lt/<  (less than), gte/>= (greater or equal), lte/<= (less or equal)
.PP
Version suffixes (e.g. 'alpha', 'rc1' etc) should use a tilde (~), e.g.: '1.2.0~alpha' to ensure proper handling
.SH OPTIONS
.TP
\fB\-D\fR, \fB\-\-debug\fR
Show debug output
.TP
\fB\-h\fR, \fB\-\-help\fR
Show this help
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Suppress all output except errors
.TP
\fB\-V\fR, \fB\-\-verbose\fR
Show verbose output
.TP
\fB\-v\fR, \fB\-\-version\fR
Show the version
.SH COPYRIGHT
Copyright \(co 2020, Koalephant Co., Ltd


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.8.1] - 2020-05-02
### Fixed
- Help output alignment in `k-tool-install`

## [2.8.0] - 2020-04-10
### Added
- `k_tool_description_add`, `k_tool_description_code_add` for adding formatted blocks of text/code to 'help' output.
- `k_tool_usage_text`, `k_tool_usage_code` for formatting blocks of test/code to use in 'help' output.
- `k_tool_environment_add` for adding a defined environment variable to 'help' output
- `k_tool_environment_alias` for adding a defined environment variable alias to 'help' output
- `k_tool_environment_print` for printing defined environment 'help' output 
- `k_function_exists` for checking if a named function is defined

### Fixed
- Allow `k_fs_resolve` to resolve paths to non-existent files/directories 

### Deprecated
- `k_tool_options_text` use `k_tool_description_add` or `k_tool_usage_text` instead
- `k_tool_options_code` use `k_tool_description_code_add` or `k_tool_usage_code` instead

## [2.7.3] - 2020-03-13
### Fixed
- Fix a bug causing `k_fs_resolve` to return some paths ending with `/`
- Fix a hang when running GPG tests with low entropy 

## [2.7.2] - 2019-11-21
### Added
- Helper named 'provision' scripts in the Vagrantfile for `build`, `test` and `test-release`

### Fixed
- Handle lack of TTY device gracefully
- Fix a bug causing `k-script-build` to hang waiting for input when there is no TTY present.
- Resolved a circular dependency issue and an overridden recipe issue in the Makefile when building temporary `k-script-build-static` asset.
- Resolved sed-related issue preventing some string-handling functions working properly on BSD/macOS systems
- Improve fallback temp file handling
- Improved memory-backed tmp file testing on systems without tmpfs filesystems
- Fix GPG tests on macOS

## [2.7.1] - 2019-11-16
### Fixed
- Let `k-tool-install` pass value of `--version-opt` to detected binaries unquoted. 

## [2.7.0] - 2019-11-15
### Added
- `k_int_{min,max,sum,avg}()` functions, to get minimum, maximum, sum and average value from a series, respectively.
- `k_string_indent()` function to indent lines of text.
- `k_string_wrap()` function to word-wrap lines of text.
- `k_string_prefix()` function to apply a prefix to lines of text.
- `k_string_get_lines()` function to retrieve multiple lines from a string
- `k_tool_options_add()` function to add an entry for formatted options help text.
- `k_tool_options_alias()` function to add an alias entry for formatted options help text.
- `k_tool_options_print()` function to print all formatted options help text.
- `k_tool_options_text()` function to print descriptive help text.
- `k_tool_options_code()` function to print example code-like text.
- `k_tool_has_options()` function to determine whether a tool accepts options.
- `k_tty_width()` function to retrieve the column width of the TTY.
 
### Changed
- `k_string_{get_line,pad_right,pad_left,lower,upper,contains,starts_with,ends_with,remove_start,remove_end,trim}()` functions can read from STDIN.
- `k_options_{split,arg_required,arg_optional}()` functions can handle option values that start with a dash by setting `$KOALEPHANT_OPTIONS_ALLOW_DASH`.
- `k_options_arg_required()` function can accept empty-string option values by setting `$KOALEPHANT_OPTIONS_ALLOW_EMPTY`.
- `k_fs_temp_{get_raw,dir,file,predictable_dir,predictable_file}()` functions accept `$KOALEPHANT_TMP_USE_MEM` and `$KOALEPHANT_TMP_REQ_MEM` to try, and require using an in-memory filesystem, respectively.

### Fixed
- Prevent building release file with `unreleased` entry in CHANGELOG.
- Fix docs issue with `k_version_helper` specifying `$0` is input.
- Spelling mistakes in various shelldoc comments.
- A heap of unlikely-but-possible shell quoting issues.
- Ensure quotes and backticks in values are escaped in `k_env_set()`

## [2.6.2] - 2019-07-27
### Fixed
- Ensure dynamically loaded elements of library still work dynamically in a static script
- issue where `k-script-build` would error when processing a script provided via STDIN

## [2.6.1] - 2019-07-26
### Fixed
- missing date for 2.6.0 release
- missing version in generated `install.sh`

## [2.6.0] - 2019-07-26
### Added
- missing Unit tests for `k_gpg_*` and `k_fs_*` functions
- `k_fs_dir_empty()` function to check if a directory is empty
- `k_fs_temp_get_raw()` function to generate temporary names and get the system temp dir using either `mktemp` or `/dev/urandom`
- a generated 'install' script for use in other projects
- `k_hash_*` functions to support hash generation and verification using modules
- `php`, `shasum` and `algosum` modules for `k_hash_*` functions
- `k_config_modules()` function to list loaded config modules
- `k-tool-install` now supports hash verification on source archives
- `k-tool-install` now supports downloading via either `curl` or `wget`, depending on which is available

### Changed
- renamed `php` and `php_ro` config modules to `php-dba` and `php-ini` respectively
- moved config module functions into separate module files
- `k_options_split()`, `k_options_arg_required()` and `k_options_arg_optional()` now use a variable `$OPTION_VALUE_SET` to set/check if an option has been provided a value
- `k_options_arg_required()` now accepts a second boolean parameter, to accept an empty string value. Defaults to false 

### Fixed
- a bug in `k_fs_resolve()` with certain symlink paths
- `k_fs_temp_*()` and `k_fs_predictable_*()` functions now use `k_fs_temp_get_raw()` rather than `mktemp` directly
- a race condition bug in `k_fs_predictable_file()` when generating the predictable file for the first time
- a situation where `k_fs_predictable_file()` returned filenames without creating them
- k-tool-install erroneously always using `--dev` option with `install-deps.sh` scripts

src/k-version-compare.sh
#!/bin/sh
# Compare two version strings
# Version: PACKAGE_VERSION
# Copyright: 2019, Koalephant Co., Ltd
# Author: Stephen Reay <stephen@koalephant.com>, Koalephant Packaging Team <packages@koalephant.com>

# Process shell source (.) statements into absolute references or inline the script
k_version_compare_tool () {

	set -eu

	. ./base.lib.sh
	. ./bool.lib.sh
	. ./fs.lib.sh
	. ./string.lib.sh
	. ./number.lib.sh

	# shellcheck disable=SC2039
	local KOALEPHANT_TOOL_DESCRIPTION KOALEPHANT_TOOL_VERSION KOALEPHANT_TOOL_ARGUMENTS \
		showVersion=false verbose=false quiet=false debug=false

	readonly KOALEPHANT_TOOL_VERSION='PACKAGE_VERSION'
	readonly KOALEPHANT_TOOL_DESCRIPTION='Compare two version strings'
	readonly KOALEPHANT_TOOL_ARGUMENTS='versionString1 versionString2 [operator=gte] '

	k_tool_usage_init () {
		# shellcheck disable=SC2119
		k_tool_description_add <<-'EOT'
			Valid operators are eq/= (equal), neq/!= (not equal), gt/> (greater than), lt/<  (less than), gte/>= (greater or equal), lte/<= (less or equal)

			Version suffixes (e.g. 'alpha', 'rc1' etc) should use a tilde (~), e.g.: '1.2.0~alpha' to ensure proper handling
		EOT
		k_tool_options_add 'debug' 'D' '' 'Show debug output'
		k_tool_options_add 'help' 'h' '' 'Show this help'
		k_tool_options_add 'quiet' 'q' '' 'Suppress all output except errors'
		k_tool_options_add 'verbose' 'V' '' 'Show verbose output'
		k_tool_options_add 'version' 'v' '' 'Show the version'
	}

	k_log_level "${KOALEPHANT_LOG_LEVEL_NOTICE}" > /dev/null

	read_options() {
		# shellcheck disable=SC2039
		local originalCount="$#"

		while k_options_split "$@"; do
			case "$OPTION_NAME" in

				(-h|--help)
					k_usage
					exit
				;;

				(-V|--verbose)
					k_log_level "${KOALEPHANT_LOG_LEVEL_INFO}" > /dev/null
					verbose=true
					shift
				;;

				(-D|--debug)
					k_log_level "${KOALEPHANT_LOG_LEVEL_DEBUG}" > /dev/null
					# shellcheck disable=SC2034
					debug=true
					shift
				;;

				(-q|--quiet)
					k_log_level "${KOALEPHANT_LOG_LEVEL_ERR}" > /dev/null
					quiet=true
					shift
				;;

				(-v|--version)
					showVersion=true
					shift
				;;

				(--)
					shift
					break
				;;

				(-*)
					k_log_err 'Unknown option: "%s"' "${OPTION_NAME}"
					k_usage
					return 1
				;;

				(*)
					break
				;;
			esac
		done

		return $(( originalCount - $# ))
	}

	read_options "$@" || shift "$?"

	k_version_helper "${showVersion}" "${quiet}" "${verbose}" && return 0

	k_requires_args "$(k_tool_name)" "$#" 2

	k_version_compare "$@"
}

k_version_compare_tool "$@"
